k8s-reporter: Stop dumping unreachable container's info

oshoval opened the pull request:
**What this PR does / why we need it**:
In order to keep to logs cleaner and more informative,
refrain from printing the same error again again,
and also save time if the container / pod aren't
reachable (instead of running more commands).

Additionally - Fix logic of checking if the virt-handler's compute container is healthy,
not just exist.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #7799 

**Special notes for your reviewer**:

**Release note**:
```release-note
None
```

oshoval commented:
/cc @phoracek @qinqon

qinqon commented:
/lgtm

AlonaKaplan commented on tests/reporter/kubernetes.go:
```diff
@@ -1104,6 +1104,10 @@ func (r *KubernetesReporter) executeContainerCommands(virtCli kubecli.KubevirtCl
 				command, target, stdout, stderr, err,
 			)
+
+			if strings.Contains(err.Error(), "not found") {
```
Is there a reason not to use `errors.IsNotFound`?

oshoval replied:
It isn't a `IsNotFound` type, it is created by
`fmt.Errorf("unable to upgrade connection: %s", responseError)`
the `not found` is a string that appears in both cases of container and pod not found (see the full lines please in the ref issue)

AlonaKaplan replied:
What worries me is that we may catch another error with `not found` that is not one of the two this PR wants to catch.
On the other side, if the messages we want to catch will be changed (for example to "Not Found") the condition will stop working.

To solve at least the first issue let's use the full two messages.

oshoval replied:
[Changes](https://github.com/kubevirt/kubevirt/pull/7815#issuecomment-1142236264)

oshoval commented:
Addressed comments by checking that the pod / container are healthy,
fixed another place that checked only that the compute container exists, but didn't check it is healthy

oshoval commented:
/retest

oshoval commented:
Rebased

xpivarc commented:
Is this ready?

oshoval commented:
> Is this ready?

Basically yes, i want to do few more tests, once it will be ready i will undraft

oshoval commented:
Update inner `pod` variable to not override the outer one, so if there is a Get error, the outer won't be affected,
as it is still being used

AlonaKaplan commented on tests/reporter/kubernetes.go:
```diff
@@ -1104,6 +1106,11 @@ func (r *KubernetesReporter) executeContainerCommands(virtCli kubecli.KubevirtCl
 			)
+
+			pod, err := virtCli.CoreV1().Pods(pod.ObjectMeta.Namespace).Get(context.Background(), pod.ObjectMeta.Name, metav1.GetOptions{})
+			if errors.IsNotFound(err) || (err == nil && (pod.Status.Phase != "Running" || !isContainerReady(pod.Status.ContainerStatuses, container))) {
```
Did you tell me the error is not of kind `errors.IsNotFound`?

Why not just - `if err != nil && (pod.Status.Phase != "Running" || !isContainerReady(pod.Status.ContainerStatuses, container)))`?

oshoval replied:
> Did you tell me the error is not of kind errors.IsNotFound?
it is not the same API call, now we are doing a client.Get, before that the error was the Execute on pod

about the 2nd question
if err != nil, you can't look on the result even, it is undefined / potential not exists at all, anyhow the result is not valid as the Get failed (hence err != nil)

AlonaKaplan commented:
/approve

oshoval commented:
Thanks Alona

@qinqon can you please review again ?

oshoval commented:
/retest

oshoval commented:
/cc @EdDev

qinqon commented:
/lgtm

oshoval commented:
/retest
